### PR TITLE
Support StakeableLock(in, out).

### PR DESCRIPTION
**Pull request by jonored**
This provides support for StakeableLockIn and StakeableLockOut in the ledger app; StakeableLockIn is simply accepted as-is, and this introduces a new prompt for the user to validate on every StakeableLockOut. This is so that a compromised host cannot silently lock a user's funds, but does put validation of the date on the user. This also may produce a large number of prompts, if many different lock times are involved.

### CODE REVIEW

**Comment by ApolloUnicorn** [comment]
Closes https://github.com/obsidiansystems/ledger-app-avalanche/issues/39

File: src/parser.c [hunk 1]
```diff
@@ -328,28 +381,47 @@ enum parse_rv parse_Output(struct Output_state *const state, parser_meta_state_t
...
                     INIT_SUBPARSER(secp256k1TransferOutput, SECP256K1TransferOutput);
+                    break;
+                case 0x00000016:
+                    if(state->noStakeableLock) REJECT("Can't parse nested StakeableLockOutput");
+                    INIT_SUBPARSER(stakeableLockOutput, StakeableLockOutput);
+                    break;
             }
         case 1:
             switch (state->type) {
                 default: REJECT("Unrecognized ouput type");
```
Comment (alexfmpe) [inline src/parser.c:398]: "output". Same above

File: src/parser.c [hunk 2]
```diff
@@ -407,8 +483,31 @@ enum parse_rv parse_SECP256K1TransferInput(struct SECP256K1TransferInput_state *
...
+
+enum parse_rv parse_StakeableLockInput(struct StakeableLockInput_state *const state, parser_meta_state_t *const meta){ 
+    enum parse_rv sub_rv = PARSE_RV_INVALID;
+    switch (state->state) {
+      case 0: // Locktime
+        CALL_SUBPARSER(uint64State, uint64_t);
+        PRINTF("StakeableLockInput locktime: %.*h\n", 8, state->uint64State.buf);
+        state->state++;
+        INIT_SUBPARSER(inputState, Locked_TransferableInput);
+      case 1: // nested TransferrableOutput
```
Comment (alexfmpe) [inline src/parser.c:502]: Input?

File: src/parser.c [hunk 3]
```diff
@@ -956,12 +1077,14 @@ enum parse_rv parse_AddValidatorTransaction(struct AddValidatorTransactionState
...
             meta->swap_output = true;
+            PRINTF("DBG: Stake slot\n");
             CALL_SUBPARSER(outputsState, TransferableOutputs);
+            PRINTF("DBG: Stake slot done\n");
             state->state++;
             INIT_SUBPARSER(ownersState, SECP256K1OutputOwners);
         }
         case 2: {
-            if ( meta->staking_weight != meta->staked ) REJECT("Stake total did not match sum of stake UTXOs");
+            if ( meta->staking_weight != meta->staked ) REJECT("Stake total did not match sum of stake UTXOs: %.*h %.*h", 8, &meta->staking_weight, 8, &meta->staked);
```
[the remark this answers is not in the document]
Comment (jonored) [reply src/parser.c:1087]: Oh, we could just revert this part, I was just debugging. We could also make it clearer, and worth remembering that the "user" here had to install usb debugging to get to this message (at least for the time being) so probably has access to the source.

File: src/parser.c [hunk 4]
```diff
@@ -956,12 +1077,14 @@ enum parse_rv parse_AddValidatorTransaction(struct AddValidatorTransactionState
 
         case 1: {// Value
             meta->swap_output = true;
+            PRINTF("DBG: Stake slot\n");
```
[the remark this answers is not in the document]
Comment (jonored) [reply src/parser.c:1080]: Nope.

File: src/parser.c [hunk 5]
```diff
@@ -317,8 +320,58 @@ enum parse_rv parse_SECP256K1OutputOwners(struct SECP256K1OutputOwners_state *co
...
+        PRINTF("StakeableLockOutput locktime: %.*h\n", 8, state->uint64State.buf);
+        state->state++;
+        INIT_SUBPARSER(outputState, Locked_TransferableOutput);
+      case 1: // nested TransferrableOutput
+        CALL_SUBPARSER(outputState, Locked_TransferableOutput); // Expanded call, we need to cast here.
+        locked_prompt_t promptData;
+        promptData.amount=meta->last_output_amount;
+        promptData.until=state->locktime;
+        state->state++;
+        if( ADD_PROMPT("Funds locked", &promptData, sizeof(locked_prompt_t), lockedFundsPrompt) ) {
```
[the remark this answers is not in the document]
Comment (jonored) [reply src/parser.c:362]: fine with me; sizeof(type) was used elsewhere so I took that option.

File: src/parser.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexfmpe) [reply src/parser.c:1087]: Ah forgot only the error code is normally seen outside of debugging

File: src/parser.c (hunk 2 above)
Comment (jonored) [reply src/parser.c:502]: Yep, whoops.

File: src/parser.c (hunk 1 above)
Comment (jonored) [reply src/parser.c:398]: Huh. Not actually part of the diff, looks like; regardless, might as well fix.

**Pull request merged**